src/js.rs

```diff
@@ -11,6 +11,85 @@ impl Default for JsGenerator {
     }
 }
 
+#[derive(Default)]
+pub struct TsGenerator;
+
+impl TsGenerator {
+    pub fn generate(&self, iface: Interface) -> js::Tokens {
+        // TODO: Box Type
+        quote! {
+            /* tslint:disable */
+            /* eslint:disable */
+            export class Api {
+              constructor();
+
+              static fetch(url, imports): Promise<void>;
+
+              #(for func in iface.functions() join (#<line>#<line>) => #(self.generate_function(func)))
+            }
+
+            #(for obj in iface.objects() => #(self.generate_object(obj)))
+        }
+    }
+
+    fn generate_function(&self, func: AbiFunction) -> js::Tokens {
+        let r#static = if func.is_static {
+            quote!(static #<space>)
+        } else {
+            quote!()
+        };
+        let wrap_in_promise_begin = if func.is_async {
+            quote!(Promise<)
+        } else {
+            quote!()
+        };
+        // TODO: nested Promises
+        let wrap_in_promise_end = if func.is_async { quote!(>) } else { quote!() };
+        let args = quote!(#(for (name, ty) in &func.args join (, ) => #name: #(&wrap_in_promise_begin)#(self.generate_return_type(Some(ty)))#(&wrap_in_promise_end)));
+        quote! {
+            #r#static#(&func.name)(#args): #(self.generate_return_type(func.ret.as_ref()));
+        }
+    }
+    fn generate_return_type(&self, ret: Option<&AbiType>) -> js::Tokens {
+        if let Some(ret) = ret {
+            match ret {
+                AbiType::Prim(prim) => match prim {
+                    PrimType::U8
+                    | PrimType::U16
+                    | PrimType::U32
+                    | PrimType::Usize
+                    | PrimType::I8
+                    | PrimType::I16
+                    | PrimType::I32
+                    | PrimType::Isize
+                    | PrimType::F32
+                    | PrimType::F64 => quote!(number),
+                    PrimType::U64 | PrimType::I64 => quote!(BigInt),
+                    PrimType::Bool => quote!(boolean),
+                },
+                AbiType::RefStr | AbiType::String => quote!(string),
+                AbiType::RefSlice(prim) | AbiType::Vec(prim) => {
+                    // TODO String etcs
+                    quote!(Array<#(&self.generate_return_type(Some(&AbiType::Prim(*prim))))>)
+                }
+                AbiType::Box(i) | AbiType::Ref(i) | AbiType::Object(i) => quote!(#(i)),
+            }
+        } else {
+            quote!(void)
+        }
+    }
+    fn generate_object(&self, obj: AbiObject) -> js::Tokens {
+        quote! {
+            export class #(&obj.name) {
+                constructor(api: Api, box: Box);
```

Comment:
Not sure about the Box Type here. I guess we need a way to annotate custom constructors, or just use `new` as a convention?

Reply:
so since objects need access to the wasm instance, the constructor would we declared on a different object. For example `api.create()` would construct a new `CustomType`.